### ISSUE TITLE
on_failure_callback call multiple times and when task has retries

### ISSUE THREAD

**Issue by eejbyfeldt** [apache/airflow#30146]
### Apache Airflow version

2.5.2

### What happened

The changes in https://github.com/apache/airflow/pull/29743 adds a new places where the `on_failure_callback` is called. This leads to two incorrect behaviors. 

1. The `on_failure_callback` is incorrectly called when a task has retries and goes in `UP_FOR_RETRY`

2. The `on_failure_callback` is sometimes called twice

### What you think should happen instead

The on_failure_callback should only be called once when the task goes into a failed state.

### How to reproduce

These two patches (https://github.com/eejbyfeldt/airflow/commit/b0e7a0ae3b2c494bb75772866466110c6b3b7e8f, https://github.com/eejbyfeldt/airflow/commit/c48ca448ac3419d7b2d840405ed0b4699b8ccc02) modifies and existing test case to show that it now in correctly and the second one adds a test case showing it now gets called more than once.
```
From b0e7a0ae3b2c494bb75772866466110c6b3b7e8f Mon Sep 17 00:00:00 2001
From: Emil Ejbyfeldt <eejbyfeldt@liveintent.com>
Date: Thu, 16 Mar 2023 14:46:04 +0100
Subject: [PATCH 1/2] Modify spec to show that callback is now incorrectly
 called

---
 tests/models/test_taskinstance.py | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/tests/models/test_taskinstance.py b/tests/models/test_taskinstance.py
index 50cac05296..d4a0328513 100644
--- a/tests/models/test_taskinstance.py
+++ b/tests/models/test_taskinstance.py
@@ -448,7 +448,7 @@ class TestTaskInstance:
         ti.run()
         assert State.SKIPPED == ti.state
 
-    def test_task_sigterm_works_with_retries(self, dag_maker):
+    def test_task_sigterm_works_with_retries(self, dag_maker, caplog):
         """
         Test that ensures that tasks are retried when they receive sigterm
         """
@@ -462,6 +462,7 @@ class TestTaskInstance:
                 python_callable=task_function,
                 retries=1,
                 retry_delay=datetime.timedelta(seconds=2),
+                on_failure_callback=lambda context: context["ti"].log.info("on_failure_callback called"),
             )
 
         dr = dag_maker.create_dagrun()
@@ -471,6 +472,7 @@ class TestTaskInstance:
             ti.run()
         ti.refresh_from_db()
         assert ti.state == State.UP_FOR_RETRY
+        assert "on_failure_callback called" not in caplog.text
 
     def test_task_sigterm_calls_on_failure_callack(self, dag_maker, caplog):
         """
-- 
2.39.2
```
```
From c48ca448ac3419d7b2d840405ed0b4699b8ccc02 Mon Sep 17 00:00:00 2001
From: Emil Ejbyfeldt <eejbyfeldt@liveintent.com>
Date: Thu, 16 Mar 2023 16:10:53 +0100
Subject: [PATCH 2/2] Add test case for on_failure_callback only being called
 once

---
 tests/models/test_taskinstance.py | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/tests/models/test_taskinstance.py b/tests/models/test_taskinstance.py
index d4a0328513..8d7d8f4861 100644
--- a/tests/models/test_taskinstance.py
+++ b/tests/models/test_taskinstance.py
@@ -474,6 +474,32 @@ class TestTaskInstance:
         assert ti.state == State.UP_FOR_RETRY
         assert "on_failure_callback called" not in caplog.text
 
+    def test_task_sigterm_call_on_failure_callback_only_once(self, dag_maker, caplog):
+        """
+        Test that ensures on_failure_callback is called once on sigterm
+        """
+
+        def task_function(ti):
+            os.kill(ti.pid, signal.SIGTERM)
+
+        with dag_maker("test_mark_failure_2"):
+            task = PythonOperator(
+                task_id="test_on_failure",
+                python_callable=task_function,
+                retries=0,
+                retry_delay=datetime.timedelta(seconds=2),
+                on_failure_callback=lambda context: context["ti"].log.info("on_failure_callback called"),
+            )
+
+        dr = dag_maker.create_dagrun()
+        ti = dr.task_instances[0]
+        ti.task = task
+        with pytest.raises(AirflowException):
+            ti.run()
+        ti.refresh_from_db()
+        assert ti.state == State.FAILED
+        assert caplog.text.count("on_failure_callback called") == 1
+
     def test_task_sigterm_calls_on_failure_callack(self, dag_maker, caplog):
         """
         Test that ensures that tasks call on_failure_callback when they receive sigterm
-- 
2.39.2

```

Reverting the code changes from https://github.com/apache/airflow/pull/29743  both of these specs passes and the new spec add in that PR also succeeds without the code changes in it. So it not clear it solves the bug it intended to solve. 

### Operating System

Fedora 37

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Looks legit. cc: @ephraimbuddy

**Comment by ephraimbuddy** [comment]
I think the 'miss' is on differentiating sigterm from normal task failure 🤦‍♂️

**Comment by ephraimbuddy** [comment]
@eejbyfeldt , I will like you to test this:

```diff

diff --git a/airflow/models/taskinstance.py b/airflow/models/taskinstance.py
index ffcdfb5651..180e009de5 100644
--- a/airflow/models/taskinstance.py
+++ b/airflow/models/taskinstance.py
@@ -1580,9 +1580,12 @@ class TaskInstance(Base, LoggingMixin):
                     self.task.post_execute(context=context, result=result)
                 except AirflowTermSignal:
                     self.task.on_kill()
+                    if self.task.retries:
+                        raise AirflowException("Task received SIGTERM signal")
                     if self.task.on_failure_callback:
                         self._run_finished_callback(self.task.on_failure_callback, context, "on_failure")
-                    raise AirflowException("Task received SIGTERM signal")
+
+
 
         Stats.incr(f"operator_successes_{self.task.task_type}", 1, 1)
         Stats.incr(
diff --git a/tests/models/test_taskinstance.py b/tests/models/test_taskinstance.py
index 50cac05296..d420ab694a 100644
--- a/tests/models/test_taskinstance.py
+++ b/tests/models/test_taskinstance.py
@@ -490,8 +490,7 @@ class TestTaskInstance:
         dr = dag_maker.create_dagrun()
         ti = dr.task_instances[0]
         ti.task = task_
-        with pytest.raises(AirflowException):
-            ti.run()
+        ti.run()
         assert "on_failure_callback called" in caplog.text
 
     @pytest.mark.parametrize("state", [State.SUCCESS, State.FAILED, State.SKIPPED])
```

**Comment by eejbyfeldt** [comment]
That change does not look correct to me. The current condition that determines if the on_failure callback is called is
https://github.com/apache/airflow/blob/699171679a79917e9c2ee798f71a01d3051f3cc4/airflow/models/taskinstance.py#L1895
So the correct condition for deciding the callback would need to replicate that logic? 

Also not reraising the exception seems very problematic to me. The reason it currently called twice is that after reraiseing we end up here
https://github.com/apache/airflow/blob/699171679a79917e9c2ee798f71a01d3051f3cc4/airflow/models/taskinstance.py#L1465-L1479
which handle updating the state and running the failure fallback. Just swallowing the exception like in the patch you posted will result in the task being incorrectly marked as success.

Seems to me the reason for us not calling the `on_failure_callback` on cases where the task was externally marked as success/failed is this branch:
https://github.com/apache/airflow/blob/699171679a79917e9c2ee798f71a01d3051f3cc4/airflow/models/taskinstance.py#L1468-L1475
so it seems like the correct fix would be to call the callbacks from there. Because that is the bug that PR was aiming to solve right? Or did I misunderstand the goal?

It seems incorrect to me that assuming SIGTERM is the same thing being marked failed externally. As SIGTERM can be received for other reason, like the pod-preemption when running on k8s, and for such cases we want to go through the normal failure handling code.

**Comment by ephraimbuddy** [comment]
On further debugging, I realized why the failure callback was not being called when tasks are terminated externally. It's because the state of the task instance was already marked as failed so it runs this: https://github.com/apache/airflow/blob/b65dbaaf3f21ea5396da121bbfa7f895d0ab8516/airflow/models/taskinstance.py#L1471-L1475 instead of https://github.com/apache/airflow/blob/b65dbaaf3f21ea5396da121bbfa7f895d0ab8516/airflow/models/taskinstance.py#L1476-L1479

We can use flag to signify when we have a sigterm but it needs testing to ensure the behaviour is desired due to this:
> It seems incorrect to me that assuming SIGTERM is the same thing being marked failed externally. As SIGTERM can be received for other reason, like the pod-preemption when running on k8s, and for such cases we want to go through the normal failure handling code.

Here's my implementation which you can test and give me feedback:

```diff
commit d9b03d60323381d97e6df55d28c309ec1ae3962b
Author: Ephraim Anierobi <splendidzigy24@gmail.com>
Date:   Fri Mar 17 08:21:01 2023 +0100

    Fix failure callback when task is terminated externally
    
    The previous fix caused the callbacks to be called even on task retries.
    On debugging, I found that the AirflowException was actually caught but because the task instance
    was already in a failed state, the failure handler was not called.
     To fix this, I introduced a flag to signify that sigterm occurred so that it's handled
    properly.

diff --git a/airflow/exceptions.py b/airflow/exceptions.py
index e6ef9bd4e1..4bf946fd8e 100644
--- a/airflow/exceptions.py
+++ b/airflow/exceptions.py
@@ -29,12 +29,6 @@ if TYPE_CHECKING:
     from airflow.models import DagRun
 
 
-class AirflowTermSignal(Exception):
-    """Raise when we receive a TERM signal"""
-
-    status_code = HTTPStatus.INTERNAL_SERVER_ERROR
-
-
 class AirflowException(Exception):
     """
     Base class for all Airflow's errors.
diff --git a/airflow/models/taskinstance.py b/airflow/models/taskinstance.py
index ffcdfb5651..ae65d434db 100644
--- a/airflow/models/taskinstance.py
+++ b/airflow/models/taskinstance.py
@@ -77,7 +77,6 @@ from airflow.exceptions import (
     AirflowSensorTimeout,
     AirflowSkipException,
     AirflowTaskTimeout,
-    AirflowTermSignal,
     DagRunNotFound,
     RemovedInAirflow3Warning,
     TaskDeferralError,
@@ -539,6 +538,8 @@ class TaskInstance(Base, LoggingMixin):
         self.raw = False
         # can be changed when calling 'run'
         self.test_mode = False
+        # signify task received sigterm
+        self.received_sigterm = False
 
     @staticmethod
     def insert_mapping(run_id: str, task: Operator, map_index: int) -> dict[str, Any]:
@@ -567,6 +568,7 @@ class TaskInstance(Base, LoggingMixin):
         # correctly config the ti log
         self._log = logging.getLogger("airflow.task")
         self.test_mode = False  # can be changed when calling 'run'
+        self.received_sigterm = False  # signify task received sigterm
 
     @hybrid_property
     def try_number(self):
@@ -1468,7 +1470,7 @@ class TaskInstance(Base, LoggingMixin):
             # for case when task is marked as success/failed externally
             # or dagrun timed out and task is marked as skipped
             # current behavior doesn't hit the callbacks
-            if self.state in State.finished:
+            if self.state in State.finished and not self.received_sigterm:
                 self.clear_next_method_args()
                 session.merge(self)
                 session.commit()
@@ -1535,7 +1537,9 @@ class TaskInstance(Base, LoggingMixin):
                 os._exit(1)
                 return
             self.log.error("Received SIGTERM. Terminating subprocesses.")
-            raise AirflowTermSignal("Task received SIGTERM signal")
+            self.task.on_kill()
+            self.received_sigterm = True
+            raise AirflowException("Task received SIGTERM signal")
 
         signal.signal(signal.SIGTERM, signal_handler)
 
@@ -1574,15 +1578,9 @@ class TaskInstance(Base, LoggingMixin):
 
             # Execute the task
             with set_current_context(context):
-                try:
-                    result = self._execute_task(context, task_orig)
-                    # Run post_execute callback
-                    self.task.post_execute(context=context, result=result)
-                except AirflowTermSignal:
-                    self.task.on_kill()
-                    if self.task.on_failure_callback:
-                        self._run_finished_callback(self.task.on_failure_callback, context, "on_failure")
-                    raise AirflowException("Task received SIGTERM signal")
+                result = self._execute_task(context, task_orig)
+                # Run post_execute callback
+                self.task.post_execute(context=context, result=result)
 
         Stats.incr(f"operator_successes_{self.task.task_type}", 1, 1)
         Stats.incr(
diff --git a/tests/models/test_taskinstance.py b/tests/models/test_taskinstance.py
index 50cac05296..b9774aaeff 100644
--- a/tests/models/test_taskinstance.py
+++ b/tests/models/test_taskinstance.py
@@ -518,6 +518,32 @@ class TestTaskInstance:
         ti.refresh_from_db()
         ti.state == state
 
+    def test_task_sigterm_call_on_failure_callback_only_once(self, dag_maker, caplog):
+        """
+        Test that ensures on_failure_callback is called once on sigterm
+        """
+
+        def task_function(ti):
+            os.kill(ti.pid, signal.SIGTERM)
+
+        with dag_maker("test_mark_failure_2"):
+            task_ = PythonOperator(
+                task_id="test_on_failure",
+                python_callable=task_function,
+                retries=0,
+                retry_delay=datetime.timedelta(seconds=2),
+                on_failure_callback=lambda context: context["ti"].log.info("on_failure_callback called"),
+            )
+
+        dr = dag_maker.create_dagrun()
+        ti = dr.task_instances[0]
+        ti.task = task_
+        with pytest.raises(AirflowException):
+            ti.run()
+        ti.refresh_from_db()
+        assert ti.state == State.FAILED
+        assert caplog.text.count("on_failure_callback called") == 1
+
     @pytest.mark.parametrize(
         "state, exception, retries",
         [
```

What I did was to revert the previous commit and introduced a flag

**Comment by potiuk** [comment]
I do not have a lot of time to look at this in detail but from the quick view and speed with which we got the report (thanks @eejbyfeldt ) I think we should revert it for 2.5.3. And think about potentially improving the callback mechanism in 2.6 or beyond. Part of discussion on the new SLA proposal  (https://lists.apache.org/thread/kn1rb9l70jk16j83fycp68l2g70wnhzl) is about potential of making changes for all the callbacks (including the proposed sla callback) following much more "at least one" [delivery semantics](https://wiki.cantara.no/display/architecture/At-least+once+semantics) in a more "systematic" way - because we have loosly "at-most-once" now it seems coupled with some buggy cases where we have "not-even-once" semantics :D. 

Also dragging attention of @pierrejeambrun and @jedcunningham.

We already have quite a few fixes merged for 2.5.3 it looks, so I would be personally (also following our "release-early/release-often" discussion) be for reverting that one and releasing 2.5.3 - maybe even in a week or so. Possibly with some more scrutiny and attempts to better diagnose some of those issues that we've been regularly moving from milestone-to-milestone - there are a few of those - likely important - that seem to be perpetually in the "next" bugfix release.  Happy to join the efforts :).
